concent_api/common/decorators.py

```diff
@@ -298,3 +300,36 @@ def wrapper(*args, **kwargs):
             )
             raise
     return wrapper
+
+
+def retry_on_database_error(func):
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        repeat_handler = RepeatHandler()
+        while True:
+            try:
+                result = func(*args, **kwargs)
+                break
+            except OperationalError:
+                repeat_handler.validate_if_max_number_of_tries_exceeded()
```

Comment:
I checked it. I previous PR I used DatabaseError and You asked about it:

"Are you absolutely sure that 'DatabaseError' always indicates a failure due to a timed out lock? Because I think that there are a lot more things that can cause this error and blindly retrying when they happen might actually be harmful.

We want to retry only if it's a problem with the lock. If there's a different problem we want to raise an exception and exit immediately."

Trying to access locked row raises `OperationalError` so i did it like this, using `OperationalError`, which is `DatabaseError` subclass

Reply:
I asked about it because both `OperationalError` and `DatabaseError` do not sound like "lock timed out error". They almost certainly represent a wider class of errors. It stands out immediately.

The fact that this error is raised does not mean that it's not raised in any other case. `OperationalError` can for example be raised if the table does not exist. Deadlock is also reported as `OperationalError` and that's definitely not a situation in which we want to keep trying to lock the table. You're putting the decorator around a big piece of code (the whole Celery task) and there could be many different problems that your code would mistakenly interpret as a timed out lock.

The exception itself may not be enough to determine the cause. Maybe it contains more information about the problem? E.g. an error code from postgresql? The message looks like this: `(1213, 'Deadlock found when trying to get lock; try restarting transaction')`. `1213` looks like an error code and I think you should be able to access it on the exception object (please inspect the exception in debugger and see what's in `args`).

Also what about a situation in which we later add another lock in the task? For example any SCI operation puts a lock on a table that contains `nonce`. This could raise `OperationalError` too. You could successfully go through one lock but then fail on the second one. And your decorator would retry both pieces of code which would result in the operations done inside the first lock being repeated twice.

When catching exceptions you should put the handler around the smallest block possible. `select_for_update()` should be the only thing in the `try ... except` block. That's why a decorator is not a good place to put that block in.

Reply:
BTW, you can quote by using `>` character.

This:
```
> Are you absolutely sure that 'DatabaseError' always indicates a failure due to a timed out lock? Because I think that there are a lot more things that can cause this error and blindly retrying when they happen might actually be harmful.
>
> We want to retry only if it's a problem with the lock. If there's a different problem we want to raise an exception and exit immediately."
```

will look like this:
> Are you absolutely sure that 'DatabaseError' always indicates a failure due to a timed out lock? Because I think that there are a lot more things that can cause this error and blindly retrying when they happen might actually be harmful.
>
> We want to retry only if it's a problem with the lock. If there's a different problem we want to raise an exception and exit immediately."

Reply:
When reading https://docs.djangoproject.com/en/2.0/ref/models/querysets/#select-for-update I understood that `DatabaseError` is only for this case (but I agree the name was suspicious). It is bad by Django to not create a separate exception only for that case. I generally think that `select-for-update` is implemented there only to give developers 'anything' but not really something good.

At the end, we may implement locking on our own by raw sql queries instead of using limited framework functionality.

Reply:
`DatabaseError` is like base-class for Exceptions in Databases. Look:
https://docs.djangoproject.com/en/2.0/_modules/django/db/utils/#DatabaseError
`OperationalError` is subclass of `DatabaseError` and it is raised in case of blocked row. It has also attribute `__cause__`:
According to https://docs.djangoproject.com/en/2.0/ref/exceptions/
`As per PEP 3134, a __cause__ attribute is set with the original (underlying) database exception, allowing access to any additional information provided.`
In my case it is:
`BŁĄÐ: nie można nałożyć blokady na rekord w relacji "core_subtask"` - this is what my database 'produces'. It is possible to compare __cause__ with some string or something like this but i dont think it is a good idea.